### DISCUSSION TITLE
Semi is not a pure formatting rule

### DISCUSSION THREAD

**Discussion by my-tien** [typescript-eslint/typescript-eslint#7097] [RFCs]
### RFC

The semi rule page has this big red banner:

> DANGER
We strongly recommend you do not use this rule or any other formatting linter rules. Use a separate dedicated formatter instead. See [What About Formatting?](https://typescript-eslint.io/linting/troubleshooting/formatting) for more information.

On the linked page formatting rules are described as follows:

>Formatting: Stylistic rules that care only about trivia (semicolons, whitespace, etc.) **without impacting the runtime behavior of the code**. These rules conflict with dedicated formatters such as Prettier.

However, as far as I understand, semicolons don't fall into this category since absence of a semicolon can change the behaviour of your code. Example:

```
const obj = {}
(function () { })()
```

I know that there is an eslint rule `no-unexpected-multiline`. But I am not sure if this also works for Typescript reliably. After all, ~typescript-eslint~`typescript-eslint/semi` was introduced specifically to make the semicolon rule work for Typescript.

As long as this isn't settled, I thing it's incorrect to call `semi` a formatting rule.

### Additional Info

Any additional info...

### Before you submit your RFC, please confirm the following. If any of these required steps are not taken, we may not be able to review your RFC. Help us to help you!

- [X] I have [read the discussions guidelines](https://typescript-eslint.io/contributing/discussions) and per those guidelines, this fits a discussion category, _not_ a help request or standard issue.
- [X] I have [searched for related discussions](https://github.com/typescript-eslint/typescript-eslint/discussions) and [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

**Comment by JoshuaKGoldberg** [comment]
> semicolons don't fall into this category since absence of a semicolon can change the behaviour of your code.

True - technically, all sorts of seemingly safe changes can cause changes to code. Reordering imports  can cause changes when modules have side effects upon import; reordering object property destructures can cause changes when properties have getters that cause side effects. But the _vast, vast majority_ of uses for the `semi` rule come from folks enforcing the formatting concern. Thus, it's a formatting rule.

In other words: there can be _stylistic_ or even _logical_ **reasons** why a _formatting_ concern is valid, but the rule really does enforce the _formatting_ concern.

> I know that there is an eslint rule `no-unexpected-multiline`. But I am not sure if this also works for Typescript reliably.

Have you tried it? It does flag this case: https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MYewdgzgLgBCBGArGBeGBvAvgKABQDMBXMYKAS3BlwEoMZNqbsg&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXRADwAdEBjAF0QBNCBbBWgS3ndPSkWmj5okcGAC+IMUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

> typescript-eslint was introduced specifically to make the semicolon rule work for Typescript

This sentence caused some ... feelings in me. I will have to take some time to fully process them. 🙈

**Comment by my-tien** [comment]
Sorry, what I meant to write was this:

~typescript-eslint~`typescript-eslint/semi` was introduced specifically to make the semicolon rule work for Typescript.

So I assume there are some Typescript-specific features that ESLint doesn’t know anything about and so `no-unexpected-multiline` wouldn’t know either. But I am not very experienced with Typescript, so I don’t know which (edge-)cases exist that would need testing.

And also it was my understanding that the primary reason for a semicolon rule is not formatting but preventing programming pitfalls. At least that’s how I read the rational on the [eslint/semi](https://eslint.org/docs/latest/rules/semi) rule.

Import ordering on the other hand – while it can have an effect on the code – is in my understanding mainly done to have a tidier file and to find lines more quickly. So the **intent** concerns purely formatting reasons

**Reply by bradzacher** [reply]
> was introduced specifically to make the semicolon rule work for Typescript

It was introduced to add support for syntax locations that TS introduces that do not exist in standard TS. That is all our extension rule is for.

> And also it was my understanding that the primary reason for a semicolon rule is not formatting but preventing programming pitfalls.

The intent of the rule is to be explicit with your semicolons so that you do not rely on ASI at runtime which has a few footguns.

It's both a stylistic concern because semicolons are optional and a correctness concern because ASI has some edge cases (eg the case you presented with a line leading with a `(` and return statement with its expression on the next line).

Explicit semis are good style because it allows the dev to explicitly encode intent into the code to prevent ambiguity.

**Reply by bradzacher** [reply]
> So I assume there are some Typescript-specific features that ESLint doesn’t know anything about and so no-unexpected-multiline wouldn’t know either.

TS does not add any features that impact this rule. Most new TS syntax that matters is either declarations or expressions and the rule only really cares about the latter and handles them.

**Comment by my-tien** [comment]
Thanks for the clarification on the `no-unexpected-multiline` rule. This was my main concern for raising the question.  It seems dangerous to me to recommend not using `typescript-eslint/semi` without further information. But if the TS syntax is also covered by `no-unexpected-multiline` that’s a nice alternative. Maybe this could be mentioned on the rule page of `typescript-eslint/semi`?

Still the categorization into formatting/stylistic/logical linters is hard for me to understand.
@JoshuaKGoldberg Says `semi` is mainly about formatting concerns, @bradzacher says it’s about both stylistic and correctness concerns.
And the page [What About Formatting?](https://typescript-eslint.io/linting/troubleshooting/formatting/) defines formatting slightly differently at 2 locations:

> Formatters are tools that verify and correct whitespace issues in code, such as spacing and newlines.

> Formatting: Stylistic rules that care only about trivia (semicolons, whitespace, etc.) without impacting the runtime behavior of the code.

The second definition mentions semicolons but in the same sentence says it’s about rules that don’t impact runtime behavior. In fact, the runtime impact is presented as the distinguishing feature compared to stylistic rules.

If you compare with this rule, it seems to have less impact on the runtime than the semi-rule, yet it is not a formatting rule:
[@typescript-eslint/array-type](https://typescript-eslint.io/rules/array-type)

In summary, I think `semi` has a lot more subtleties than whitespace rules. I can imagine that they are still discouraged due to technical problems when executing the linter, however I find it important to point out these subtleties and to explain if they can become problematic or not.

**Reply by bradzacher** [reply]
> JoshuaKGoldberg Says semi is mainly about formatting concerns, bradzacher says it’s about both stylistic and correctness concerns.

To clarify - I meant that it is a stylistic / formatting concern primarily, same as Josh.

I was just also pointing out why the docs talk about correctness and possible errors - because of ASI and the rare footguns with accidental multi-line statements.

> The second definition mentions semicolons but in the same sentence says it’s about rules that don’t impact runtime behavior. In fact, the runtime impact is presented as the distinguishing feature compared to stylistic rules.

The presence of semicolons does not impact runtime behaviour - as long as you correctly write code.

You cannot compare
```js
foo;
(bar);
```
And
```js
foo
(bar)
```

Because they are two different code fragments. Instead the comparison would be with
```js
foo
;(bar)
```
Which is the "no semi" approach to avoiding the footgun.

It's a stylistic choice - if you choose the no-semi style you have to write the no-semi style code and you must know and understand ASI to avoid runtime breakages.

Its the same with using if/for/while blocks or not (eg `if (cond) {code}` vs `if (cond) code`) - it's a stylistic choice and if you go for the no-block style you have to be aware of its limitations and the potential footguns.
